[SPARK-6487][MLlib] Add sequential pattern mining algorithm PrefixSpan to Spark MLlib

zhangjiajin opened the pull request:
Add parallel PrefixSpan algorithm and test file.
Support non-temporal sequences.

zhangjiajin committed:
Add new algorithm PrefixSpan and test file.

AmplabJenkins commented:
Can one of the admins verify this patch?

jackylk commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+ *       (Wikipedia)]]
+ */
+class Prefixspan(
+    val sequences: RDD[Array[Int]],
```
Can you make it generic type so that it can accept not only Array[Int] as input?

zhangjiajin replied:
The Int data type is small than some other data types, such as String etc. So, inside the algorithm, we use Int data type to get the better performance.

If user use other data type, like String, maybe he will code String to Int, and decode the result to String. The following:

    val sequences = Array(
      Array("3", "1", "3", "4", "5"),
      Array("2", "3", "1"),
      Array("3", "4", "4", "3"),
      Array("1", "3", "4", "5"),
      Array("2", "4", "1"),
      Array("6", "5", "3"))
    val rdd = sc.parallelize(sequences, 2).cache()

    // create coder and decoder
    val letterMap = rdd.flatMap(x => x.distinct).distinct().zipWithIndex().mapValues(_.toInt).collect
    val coder = letterMap.toMap
    val decoder = letterMap.map(x => (x._2, x._1)).toMap

	// code
    val intRdd = rdd.map(x => x.map(y => coder(y)))

    val prefixspan1 = new Prefixspan(intRdd, 2, 50)
    val result = prefixspan1.run()

    // decode
    val stringResult = result.map(x => (x._1.map(y => decoder(y)), x._2))

I think this is a general job, some other algorithms maybe need this function too, so can we add the code and decoder as a separate model for all other algorithms ?

mengxr replied:
`FPGrowth` accepts generic item type and encode items into indices after the first step. We can do that in a follow-up PR.

feynmanliang replied:
The problem definition in the referenced paper suggests this should actually be `Array[Array[Item]]`. I agree that this extension can be deferred to a later PR and the current code is fine as is.

feynmanliang replied:
Actually can the change `Array[Int]` -> `Array[Array[Int]]` be made in this PR since it affects the kinds exposed in a public API and will be more easy to generalize (for now we can just flatten the array and use the existing implementation)?

mengxr replied:
@feynmanliang The generalization for non-temporal items would happen in a follow-up PR before 1.5. This is discussed on the JIRA page.

feynmanliang replied:
@mengxr This PR actually implements non-temporal (by temporal you mean consecutive correct?) sequential pattern mining, see the L86 and L149. My suggestion was to make the definition of a sequence consistent with the Han et al paper (An **itemset** is a subset of items. A **sequence** is an ordered list of itemsets) since this PR currently defines a sequence to be a ordered list of singleton itemsets.

feynmanliang replied:
Synced offline with @mengxr, `Array[Int]` OK for this PR.

zhangjiajin replied:
OK, It would be fixed in a follow-up PR before 1.5. It's soon.

mengxr commented:
add to whitelist

mengxr commented:
@feynmanliang Could you take a look?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36686 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36686/consoleFull) for   PR 7258 at commit [`91fd7e6`](https://github.com/apache/spark/commit/91fd7e66d0c363e68bc9ebe2bf3e03c26ef348d2).

SparkQA commented:
[Test build #36686 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36686/console) for   PR 7258 at commit [`91fd7e6`](https://github.com/apache/spark/commit/91fd7e66d0c363e68bc9ebe2bf3e03c26ef348d2).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class Prefixspan(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+import org.apache.spark.rdd.RDD
+
+/**
+ *
```
Add `::Experimental::` here and `@Experimental` to `class PrefixSpan`.

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+ *
+ * A parallel PrefixSpan algorithm to mine sequential pattern.
+ * The PrefixSpan algorithm is described in
+ * [[http://web.engr.illinois.edu/~hanj/pdf/span01.pdf]].
```
use DOI link: http://doi.org/10.1109/ICDE.2001.914830

feynmanliang replied:
Document that this implementation only covers itemsets of size 1 (Han et al define an itemset to be a subset of all items and a sequential pattern to be an ordered sequence of itemsets).

Also document whether this implementation supports non-contiguous (subsequences which skip items from the original sequence between each item) sequential patterns since Han et al define sequential patterns to be non-contiguous subsequences (Section 2: Def of **subsequence**) and if I recall correctly from our phone conversation this implementation will only support contiguous subsequences.

zhangjiajin replied:
@mengxr Fixed the link.
@feynmanliang You are right, the code supports non-contiguous.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+ * The PrefixSpan algorithm is described in
+ * [[http://web.engr.illinois.edu/~hanj/pdf/span01.pdf]].
+ *
+ * @param sequences original sequences data
```
Move this to `run` as an input argument.

zhangjiajin replied:
Fixed

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+ * @see [[https://en.wikipedia.org/wiki/Sequential_Pattern_Mining Sequential Pattern Mining
+ *       (Wikipedia)]]
+ */
+class Prefixspan(
```
`Prefixscan` -> `PrefixScan`

zhangjiajin replied:
Fixed, change Prefixspan to PrefixSpan.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+class Prefixspan(
+    val sequences: RDD[Array[Int]],
+    val minSupport: Int = 2,
+    val maxPatternLength: Int = 50) extends java.io.Serializable {
```
Is `50` too big as the default? What is the default value SPMF and R?

feynmanliang replied:
R's [arulesSequence](http://cran.r-project.org/web/packages/arulesSequences/arulesSequences.pdf) package defaults `maxlen = 10` and `support = 0.1` (with support defined as proportion of transactions in database containing subsequence)

zhangjiajin replied:
Fixed, changed maxPatternLength to 10, changed minSupport to 0.1.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+      .flatMap(p => p)
+      .map((_, 1))
+      .reduceByKey(_ + _)
+
```
Blank lines are used to separate code blocks but not individual statements in Spark.

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+      sequence.takeRight(sequence.length - index - 1)
+    }
+  }
+}
```
newline at the end

zhangjiajin replied:
Fixed.

mengxr commented:
@zhangjiajin Thanks for making this PR minimal! I made some comments about code style. @feynmanliang will make a pass on the implementation. @jackylk Could you make a pass as well?

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+ *
+ * @param sequences original sequences data
+ * @param minSupport the minimal support level of the sequential pattern, any pattern appears
+ *                   more than minSupport times will be output
```
We should be consistent with `FPGrowth` and define `minSupport` to be a double in [0,1] indicating the proportion of transactions containing the sequential pattern (I realize that Han et al define it as the total count but [Wikipedia](https://en.wikipedia.org/wiki/Association_rule_learning#Support) follows the same definition we use).

This just means that we will need to multiply by `sequences.count()` before doing any of the filtering by `minSupport`.

zhangjiajin replied:
Fixed, change minSupport data type to double in [0,1]

feynmanliang commented on mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixspanSuite.scala:
```diff
@@ -0,0 +1,47 @@
+
+    val rdd = sc.parallelize(sequences, 2).cache()
+
+    val prefixspan1 = new Prefixspan(rdd, 2, 50)
```
R code to verify results (please include in multiline comment)

```r
library("arulesSequences")
prefixSpanSeqs = read_baskets("prefixSpanSeqs", info = c("sequenceID","eventID","SIZE"))
freqItemSeq = cspade(prefixSpanSeqs,
                     parameter = list(support = 2 / length(unique(transactionInfo(prefixSpanSeqs)$sequenceID)), maxlen = 2 ))
resSeq = as(freqItemSeq, "data.frame")
resSeq
```

zhangjiajin replied:
Fixed, verified the result.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+   * d) local PrefixSpan on each prefix
+   * @return sequential patterns
+   */
+  def run(): RDD[(Seq[Int], Int)] = {
```
`sequences`should be an argument to `run()` rather than a constructor parameter.

zhangjiajin replied:
Fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+ */
+class Prefixspan(
+    val sequences: RDD[Array[Int]],
+    val minSupport: Int = 2,
```
These should be `private var`s with public getters and setters to decouple interface from implementation (see [FPGrowth](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/fpm/FPGrowth.scala#L76))

zhangjiajin replied:
Changed these input parameters to private var s, and add getters and setters.

But I have a question, why don't we use the public input parameters ?
The following code, the class A is smaller than class B, Why is class B code style better ?

    class A (var x: Int = 12) {
      def f() = {
        println(x)
      }
    }

    class B (private var x: Int) {
      def this() = this(12)
      def setX(x: Int): this.type = {
        this.x=x
        this
      }
      def f() = {
        println(x)
      }
    }

    val a = new A()
    a.f()
    a.x = 15
    a.f()

    val b = new B()
    b.f()
    b.setX(15)
    b.f()

feynmanliang replied:
Suppose you not wanted to ensure that `x` is always `>=0`. In `B`, you would just redefine
```scala
def setX(x: Int): this.type = {
  require(x >= 0)
  ...
}
```
whereas I'm not too sure how to get that done easily in `A`.

Also, you can now chain method calls to `B`, for example:
```scala
val b = new B()
b.setX(15).f()
```

feynmanliang commented on mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixspanSuite.scala:
```diff
@@ -0,0 +1,47 @@
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+
+class PrefixspanSuite extends SparkFunSuite with MLlibTestSparkContext {
```
`PrefixspanSuite` -> `PrefixSpanSuite`

zhangjiajin replied:
Fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+   * @return length-one patterns and projection table
+   */
+  private def findPatternsLengthOne(): (RDD[(Int, Int)], RDD[(Seq[Int], Array[Int])]) = {
+    val patternsOneLength = sequences
```
nit: `patternsOneLength`->`lengthOnePatternsAndCounts`

zhangjiajin replied:
Fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+  private def findPatternsLengthOne(): (RDD[(Int, Int)], RDD[(Seq[Int], Array[Int])]) = {
+    val patternsOneLength = sequences
+      .map(_.distinct)
+      .flatMap(p => p)
```
Combine L65 and L66 into `.flatMap(_.distinct)`

zhangjiajin replied:
Fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+      .map((_, 1))
+      .reduceByKey(_ + _)
+
+    val removedElements: Array[Int] = patternsOneLength
```
nit: `removedElements` -> `infrequentLengthOnePatterns`

zhangjiajin replied:
Fixed

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+      .map(_._1)
+      .collect()
+
+    val savedElements = patternsOneLength.filter(_._2 >= minSupport)
```
nit: `savedElements` -> `frequentLengthOnePatterns`

zhangjiajin replied:
Fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+    if (index == -1) {
+      Array()
+    } else {
+      sequence.takeRight(sequence.length - index - 1)
```
nit: `sequence.drop(index+1)`

zhangjiajin replied:
Fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+   */
+  private def repartitionSequences(
+      data: RDD[(Seq[Int], Array[Int])]): RDD[(Seq[Int], Array[Array[Int]])] = {
+    val dataRemovedEmptyLine = data.filter(x => x._2.nonEmpty)
```
Why not just do the filter up on L95 when creating `prefixAndCandidates` to localize the logic?

zhangjiajin replied:
Fixed. Moved the filter to L95.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+    val dataRemovedEmptyLine = data.filter(x => x._2.nonEmpty)
+    val dataMerged = dataRemovedEmptyLine
+      .groupByKey()
+      .map(x => (x._1, x._2.toArray))
```
`.map(x => ...)` -> `.mapValues(_.toArray)`

zhangjiajin replied:
Fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+  /**
+   * Re-partition the RDD data, to get better balance and performance.
+   * @param data patterns and projected sequences data before re-partition
+   * @return patterns and projected sequences data after re-partition
```
This isn't really a performance-oriented repartition in [the sense Spark has defined repartition](http://spark.apache.org/docs/1.2.2/api/java/org/apache/spark/rdd/RDD.html#repartition(int, scala.math.Ordering)).

How about calling it `makePrefixProjectedDatabses` make it clear that this the projected database construction step in PrefixScan?

zhangjiajin replied:
Changed the function name to makePrefixProjectedDatabases. It's better than old one.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+      data: RDD[(Seq[Int], Array[Int])]): RDD[(Seq[Int], Array[Array[Int]])] = {
+    val dataRemovedEmptyLine = data.filter(x => x._2.nonEmpty)
+    val dataMerged = dataRemovedEmptyLine
+      .groupByKey()
```
@mengxr could collecting the entire projected database overload a worker (e.g. imagine if a prefix was contained in every transaction, then the entire transaction database would be grouped under that key)?

feynmanliang replied:
Synced offline with @mengxr, OK for this PR.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+   */
+  private def getPatternsWithPrefix(
+      prefix: Seq[Int],
+      data: Array[Array[Int]]): Array[(Seq[Int], Int)] = {
```
nit: `data` -> `projectedDatabase`

feynmanliang replied:
Also, this assumes that the entire `projectedDatabase` will fit on a single worker. In the worst case (e.g. if a prefix is in every transaction in the database), this assumes that the entire dataset will fit on a single worker. @mengxr

feynmanliang replied:
Synced offline with @mengxr, OK for this PR.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+  private def getPatternsWithPrefix(
+      prefix: Seq[Int],
+      data: Array[Array[Int]]): Array[(Seq[Int], Int)] = {
+    val elements = data
```
Duplicate code with `findPatternsLengthOne()`, can we refactor into a common method?

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+      val sub = cleanedSearchSpace.map(y => getSuffix(x._1, y)).filter(_.nonEmpty)
+      (prefix ++ Seq(x._1), sub)
+    }.filter(x => x._2.nonEmpty)
+      .toArray
```
L136 - 155 is very similar to `findPatternsLengthOne`, I wonder if we can refactor into common method?

zhangjiajin replied:
findPatternsLengthOne and getPatternsWithPrefix do the similar work, but the input data type is different. findPatternsLengthOne faces the entire RDD data, getPatternsWithPrefix faces part of RDD data, but the content of the data is similar. Before shuffle, the prefix and the projected database distributed in all the partitions. After shuffle, one partition only including one prefix and its projected database. In other words, after shuffle, one partition of the new RDD is same as the old entire RDD. So, findPatternsLengthOne is different from getPatternsWithPrefix, eg. findPatternsLengthOne has some collect methods.

feynmanliang replied:
Gotcha :smile:. This is fine for this PR.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+    val cleanedSearchSpace = data
+      .map(x => x.filter(y => selectedSingleElements.contains(y)))
+
+    val newSearchSpace = selectedSingleElements.map { x =>
```
nit: `newSearchSpace` -> `prefixProjectedDatabases`

zhangjiajin replied:
fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+
+    val selectedElements = selectedSingleElements
+      .map(x => (prefix ++ Seq(x._1), x._2))
+      .toArray
```
nit: `selectedElements` -> `frequentPrefixes`

zhangjiajin replied:
fixed

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala:
```diff
@@ -0,0 +1,183 @@
+      .groupBy(x => x)
+      .map(x => (x._1, x._2.length))
+
+    val selectedSingleElements = elements.filter(x => x._2 >= minSupport)
```
Only the keys of this `Map` are used later, how about calling `.keys()` here to avoid all the `._1`s later and to make logic clearer?

feynmanliang replied:
nit: `selectedSingleElements` -> `frequentPrefixExtensions`

zhangjiajin replied:
Fixed

feynmanliang commented:
That's all for now! Overall:

 * I am concerned with the scalability of grouping all the prefix-projected databases by prefix (Prefixspan.scala#109) followed by local processing. This may cause the entire dataset to be sent to a single worker (see inline comments).
     * The algorithm is not truly distributed: only the first iteration (length 1 prefixes) is performed in a distributed manner on `RDD`s; the remaining iterations are all done locally.
 * There is a lot of duplication between `getPatternsInLocal`+`getPatternsWithPrefix` and the other methods. It seems like the former two are just doing the same thing except on local `Array[Array[_]]` rather than `RDD[_]`. If we can refactor things into a common method operating on `RDD[_]`, I think we can both solve the scalability problem and DRY up the code.

zhangjiajin committed:
Modified the code according to the review comments.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36779 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36779/consoleFull) for   PR 7258 at commit [`575995f`](https://github.com/apache/spark/commit/575995f69dadad825d97f2248599eb62c1743fe7).

SparkQA commented:
[Test build #36779 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36779/console) for   PR 7258 at commit [`575995f`](https://github.com/apache/spark/commit/575995f69dadad825d97f2248599eb62c1743fe7).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PrefixSpan(`
  * `class Prefixspan(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

zhangjiajin committed:
Delete Prefixspan.scala

Use PrefixSpan.scala instead of Prefixspan.scala. Delete Prefixspan.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

zhangjiajin committed:
Delete PrefixspanSuite.scala

Use PrefixSpanSuite.scala instead of PrefixspanSuite.scala, Delete PrefixspanSuite.scala.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36782 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36782/consoleFull) for   PR 7258 at commit [`a2eb14c`](https://github.com/apache/spark/commit/a2eb14c7fb6abb70eaa046baf78da205c7a4ca7d).

SparkQA commented:
[Test build #36782 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36782/console) for   PR 7258 at commit [`a2eb14c`](https://github.com/apache/spark/commit/a2eb14c7fb6abb70eaa046baf78da205c7a4ca7d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PrefixSpan(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zhangjiajin committed:
Fixed a Scala style error.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36785 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36785/consoleFull) for   PR 7258 at commit [`89bc368`](https://github.com/apache/spark/commit/89bc368f76c40ad0090a928cec49cd9d28ce666e).

SparkQA commented:
[Test build #36785 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36785/console) for   PR 7258 at commit [`89bc368`](https://github.com/apache/spark/commit/89bc368f76c40ad0090a928cec49cd9d28ce666e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class PrefixSpan(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

zhangjiajin commented:
@feynmanliang  findPatternsLengthOne and getPatternsWithPrefix do the similar work, but the input data type is different. findPatternsLengthOne faces the entire RDD data, getPatternsWithPrefix faces part of RDD data, but the content of the data is similar. Before shuffle, the prefix and the projected database distributed in all the partitions. After shuffle, one partition only including one prefix and its projected database. In other words, after shuffle, one partition of the new RDD is same as the old entire RDD. So, findPatternsLengthOne is different from getPatternsWithPrefix, eg. findPatternsLengthOne has some collect methods.

zhangjiajin commented:
About the parallel method, I think method 2 performance is better than method 1.

![image](https://cloud.githubusercontent.com/assets/13159256/8570197/6e17a238-25b1-11e5-915d-89b5a583722e.png)

![image](https://cloud.githubusercontent.com/assets/13159256/8570252/ac3c29d0-25b1-11e5-8be9-ef726d853ea4.png)

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+ *       (Wikipedia)]]
+ */
+@Experimental
+class PrefixSpan(
```
Make the constructor package private.

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+@Experimental
+class PrefixSpan(
+    private var minSupport: Double,
+    private var maxPatternLength: Int) extends java.io.Serializable {
```
`extends Serializable with Logging`

1. `java.io` is not necessary.
2. We need `Logging` for instrumentation.

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+    private var minSupport: Double,
+    private var maxPatternLength: Int) extends java.io.Serializable {
+
+    private var absMinSupport: Int = 0
```
Fix indentation and rename it to `minCount` to make clear that `support` is in [0, 1]. `abs` might be recognized as `math.abs`, which is confusing. It shouldn't be a private variable because `sequences` is given at `run` and it should be stateless.

feynmanliang replied:
I think we should remove this and make L78 a `val` since this may change on every call to `run`

zhangjiajin replied:
Fixed, removed var absMinSupport instead of val minCount.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+   * Constructs a default instance with default parameters
+   * {minSupport: `0.1`, maxPatternLength: 10}.
+   */
+  def this() = this(0.1, 10)
```
Move the auxiliary constructor before `private var`.

zhangjiajin replied:
Fixed

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+  }
+
+  /**
+   * Sets maximal pattern length.
```
Mention default value?

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+
+  /**
+   * Calculate sequential patterns:
+   * a) find and collect length-one patterns
```
This is implementation detail, which should go inside the function block. In the API doc, we should explain the input and output. E.g., what is `sequences` and what does `(Seq[Int], Int))` mean in the output?

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+   * d) local PrefixSpan on each prefix
+   * @return sequential patterns
+   */
+  def run(sequences: RDD[Array[Int]]): RDD[(Seq[Int], Int)] = {
```
Change the output to `(Array[Int], Long)` to be more Java-friendly. In a follow-up PR, we can consider wrapping it in a case class, as in `FPGrowth.FreqItemset`.

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+    allPatterns
+  }
+
+  private def getAbsoluteMinSupport(sequences: RDD[Array[Int]]): Int = {
```
Return `Long` (or `Double`)

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+  }
+
+  private def getAbsoluteMinSupport(sequences: RDD[Array[Int]]): Int = {
+    val result = if (minSupport <= 0) {
```
Should put `minSupport` range check in `setMinSupport`. This function becomes `sequences.count() * minSupport`.

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+  }
+
+  /**
+   * Find the patterns that it's length is one
```
This method does much more than it claims. The method we discuss is collect frequent items first and then broadcast them to be used for projection.

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+   */
+  private def findLengthOnePatterns(
+      sequences: RDD[Array[Int]]): (RDD[(Int, Int)], RDD[(Seq[Int], Array[Int])]) = {
+    val LengthOnePatternAndCounts = sequences
```
`LengthOnePatternAndCounts` -> `lengthOnePatternAndCounts` (use camelCase for variable names)

This should be similar to `getFreqItems` in FPGrowth:

https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/fpm/FPGrowth.scala#L118

zhangjiajin replied:
Fixed, Add a new method.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+    val LengthOnePatternAndCounts = sequences
+      .flatMap(_.distinct.map((_, 1)))
+      .reduceByKey(_ + _)
+    val infrequentLengthOnePatterns: Array[Int] = LengthOnePatternAndCounts
```
Why do we need infrequent items?

zhangjiajin replied:
Fixed, removed infrequent items.
If infrequent items is empty, we don't need filter the sequence, maybe performance is better.
When support is small, the infrequent items always empty.

  val filteredSequences =
      if (infrequentLengthOnePatterns.isEmpty) {
        sequences
      } else {
        sequences.map { p =>
          p.filter { x => !infrequentLengthOnePatterns.contains(x) }
        }
      }

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+   */
+  private def makePrefixProjectedDatabases(
+      data: RDD[(Seq[Int], Array[Int])]): RDD[(Seq[Int], Array[Array[Int]])] = {
+    val dataMerged = data
```
`data.groupByKey().mapValues(_.toArray)` It is not necessary to declare `dataMerged`.

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+   */
+  private def getPatternsInLocal(
+      data: RDD[(Seq[Int], Array[Array[Int]])]): RDD[(Seq[Int], Int)] = {
+    val result = data.flatMap { x =>
```
Same here. It is not necessary to declare `result`.

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala:
```diff
@@ -0,0 +1,69 @@
+    def formatResultString(data: RDD[(Seq[Int], Int)]): String = {
+      data.map(x => x._1.mkString(",") + ": " + x._2)
+        .collect()
+        .sortWith(_<_)
```
`.sortWith(_<_)` -> `sorted`

zhangjiajin replied:
Fixed, removed this code.

mengxr commented on mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala:
```diff
@@ -0,0 +1,69 @@
+    val rdd = sc.parallelize(sequences, 2).cache()
+
+    def formatResultString(data: RDD[(Seq[Int], Int)]): String = {
+      data.map(x => x._1.mkString(",") + ": " + x._2)
```
Converting to strings is not necessary. We can check the values directly. We need to write more lines, but it simplifies the logic.

zhangjiajin replied:
Fixed.

mengxr commented on mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala:
```diff
@@ -0,0 +1,69 @@
+    val len1 = result1.count().toInt
+    val actualValue1 = formatResultString(result1)
+    val expectedValue1 =
+      "1,3,4,5: 2; 1,3,4: 2; 1,3,5: 2; 1,3: 2; 1,4,5: 2;" +
```
~~~
(Seq(1, 3, 4, 5), 2L), (Seq(1, 3, 4), 2L), ...
~~~

zhangjiajin replied:
Fixed, instead of Array[(Array[Int], Long)]

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+   * @return sequential patterns
+   */
+  def run(sequences: RDD[Array[Int]]): RDD[(Seq[Int], Int)] = {
+    absMinSupport = getAbsoluteMinSupport(sequences)
```
Should also check whether the input dataset is cached or not. See https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/fpm/FPGrowth.scala#L95.

zhangjiajin replied:
Fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+  }
+
+  /**
+   * Re-partition the RDD data, to get better balance and performance.
```
The current wording may be confusing since `repartition` is an actual method on `RDD`s. How about changing the scaladoc to "Constructs prefix-projected databases from (prefix, suffix) pairs".

zhangjiajin replied:
Fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+      .filter(_._2 < absMinSupport)
+      .map(_._1)
+      .collect()
+    val frequentLengthOnePatterns = LengthOnePatternAndCounts
```
`frequentLengthOnePatterns` -> `frequentLengthOnePatternAndCounts`

zhangjiajin replied:
Fixed

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,209 @@
+        sequences
+      } else {
+        sequences.map { p =>
+          p.filter { x => !infrequentLengthOnePatterns.contains(x) }
```
How about (after renaming L112) creating `val frequentLengthOnePatterns = frequentLengthOnePatternAndCounts.map(_._1).toSet` and then just doing 
```scala
val filteredSequences = sequences.map { p =>
  p.filter { frequentLengthOnePatterns.contains(_) }
}
```

zhangjiajin replied:
Fixed, used your code.
If infrequent items is empty, we don't need filter the sequence, maybe performance is better.
When support is small, the infrequent items always empty.

val filteredSequences =
if (infrequentLengthOnePatterns.isEmpty) {
sequences
} else {
sequences.map { p =>
p.filter { x => !infrequentLengthOnePatterns.contains(x) }
}
}

feynmanliang commented on mllib/src/test/scala/org/apache/spark/mllib/fpm/PrefixSpanSuite.scala:
```diff
@@ -0,0 +1,69 @@
+
+class PrefixspanSuite extends SparkFunSuite with MLlibTestSparkContext {
+
+  test("Prefixspan sequences mining using Integer type") {
```
Please include the `R` code I provided so that other users can use it to verify the results. Add a multiline comment like this
```scala
/*
   library("arulesSequences")
   prefixSpanSeqs = read_baskets("prefixSpanSeqs", info = c("sequenceID","eventID","SIZE"))
   freqItemSeq = cspade(
     prefixSpanSeqs,
     parameter = list(support = 2 / length(unique(transactionInfo(prefixSpanSeqs)$sequenceID)), maxlen = 2 ))
   resSeq = as(freqItemSeq, "data.frame")
   resSeq
 */
```

zhangjiajin replied:
Fixed.

feynmanliang commented:
If you `collect` L112's `frequentLengthOnePatternAndCounts` (after renaming), L112-129 and L169-183 are basically identical (`frequentLengthOnePatternAndCounts` <=> `frequentPrefixesAndCounts`, `filteredSequences` <=> `cleanedSearchSpace`, `prefixAndCandidates` <=> `prefixProjectedDatabases`) and both operate on the same type `Array[Int]`. Can we pull out a common helper method?

mengxr commented:
@zhangjiajin The issue with method 2 is projection before filtering. It may increase the shuffle size. After we generate possible prefixes, we should count their frequencies (that does not require shuffling the suffixes), filter out invalid candidates, and then project. Basically, I'm suggesting the following:

1. generate prefix candidates
2. count their frequencies
3. filter and collect frequent ones
4. Do we have enough candidates to distribute the work? If no, go to 1 and generate candidates with length + 1.
5. project and group by prefixes
6. generate frequent patterns for each prefix
7. union with local frequent sequences

zhangjiajin committed:
Modified the code according to the review comments.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36933 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36933/consoleFull) for   PR 7258 at commit [`1dd33ad`](https://github.com/apache/spark/commit/1dd33ad82499b9ad1b446b96f2f88519ffbe9a1b).

SparkQA commented:
[Test build #36933 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36933/console) for   PR 7258 at commit [`1dd33ad`](https://github.com/apache/spark/commit/1dd33ad82499b9ad1b446b96f2f88519ffbe9a1b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zhangjiajin committed:
Fix some Scala style errors.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36935 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36935/consoleFull) for   PR 7258 at commit [`4c60fb3`](https://github.com/apache/spark/commit/4c60fb36148206abd67fe51cea667ee3d63e490e).

SparkQA commented:
[Test build #36935 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36935/console) for   PR 7258 at commit [`4c60fb3`](https://github.com/apache/spark/commit/4c60fb36148206abd67fe51cea667ee3d63e490e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zhangjiajin committed:
Fix a Scala style error.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36936 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36936/consoleFull) for   PR 7258 at commit [`ba5df34`](https://github.com/apache/spark/commit/ba5df346543e9aee119bd781b257860b65bbe7df).

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+   * Sets the minimal support level (default: `0.1`).
+   */
+  def setMinSupport(minSupport: Double): this.type = {
+    require(minSupport >= 0 && minSupport <= 1)
```
Provide an error message with the value of `minSupport`.

zhangjiajin replied:
fixed

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+   * Sets maximal pattern length (default: `10`).
+   */
+  def setMaxPatternLength(maxPatternLength: Int): this.type = {
+    require(maxPatternLength >= 1)
```
ditto

zhangjiajin replied:
fixed

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+    if (sequences.getStorageLevel == StorageLevel.NONE) {
+      logWarning("Input data is not cached.")
+    }
+    val minCount = getAbsoluteMinSupport(sequences)
```
`getMinCount(sequences)`

zhangjiajin replied:
fixed

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+    val minCount = getAbsoluteMinSupport(sequences)
+    val (lengthOnePatternsAndCounts, prefixAndCandidates) =
+      findLengthOnePatterns(minCount, sequences)
+    val repartitionedRdd = makePrefixProjectedDatabases(prefixAndCandidates)
```
`repartitionedRdd` -> `projected`?

zhangjiajin replied:
fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+  }
+
+  /**
+   * Get the absolute minimum support value (sequences count * minSupport).
```
Please update the doc and change `absolute minimum support` to `minimum count`.

zhangjiajin replied:
fixed

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+   * @return absolute minimum support value,
+   */
+  private def getAbsoluteMinSupport(sequences: RDD[Array[Int]]): Long = {
+    if (minSupport == 0) 0L else (sequences.count() * minSupport).toLong
```
`(sequences.count() * minSupport).toLong` -> `math.ceil(sequences.count() * minSupport).toLong` (using Double is actually simpler)

zhangjiajin replied:
fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+
+  /**
+   * Generates frequent items by filtering the input data using minimal support level.
+   * @param minCount the absolute minimum support
```
ditto. `support` indicates fraction. Using the same name for count is confusing. Just to be consistent.

zhangjiajin replied:
fixed

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+      sequences: RDD[Array[Int]]): RDD[(Int, Long)] = {
+    sequences.flatMap(_.distinct.map((_, 1L)))
+      .reduceByKey(_ + _)
+      .filter(_._2 >= minCount)
```
I think you should collect it directly. When generating the final patterns, use `sc.parallelize`, which might be cheaper.

zhangjiajin replied:
fixed

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+      minCount: Long,
+      sequences: Array[Array[Int]]): Array[(Int, Long)] = {
+    sequences.flatMap(_.distinct)
+      .groupBy(x => x)
```
This is too expensive (creating many temp objects). You only need the counts, so create an PrimitiveKeyOpenHashMap (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/util/collection/PrimitiveKeyOpenHashMap.scala) and count one by one.

zhangjiajin replied:
For example, the sequences is: <abadgfg> <accede><bddbe>, the frequent item and counts is: (a:2),(b:2),(c:1),(d:3),(e:2),(f:1),(g:1). I have no idea how to use PrimitiveKeyOpenHashMap.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+      sequences: Array[Array[Int]]): Array[(Int, Long)] = {
+    sequences.flatMap(_.distinct)
+      .groupBy(x => x)
+      .mapValues(_.length.toLong)
```
Just thought about this again. Our implementation does not actually support Long (maybe except length-1 patterns), because we collect the projected database as an Array, which only support at most Int.MaxValue items. We can keep `Long` here and discuss this later.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+   * @return prefixes and projected database
+   */
+  private def getPatternAndProjectedDatabase(
+      frequentPrefixes: Array[Int],
```
use a `Set` (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala)

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+      frequentPrefixes: Array[Int],
+      sequences: RDD[Array[Int]]): RDD[(Array[Int], Array[Int])] = {
+    val filteredSequences = sequences.map { p =>
+      p.filter (frequentPrefixes.contains(_) )
```
otherwise, `contains` is very expensive (linear time).

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+        val sub = getSuffix(y, x)
+        (Array(y), sub)
+      }
+    }.filter(x => x._2.nonEmpty)
```
put the filter inside `flatMap` to reduce RDD lineage

zhangjiajin replied:
fixed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,252 @@
+  private def findLengthOnePatterns(
+      minCount: Long,
+      sequences: RDD[Array[Int]]): (RDD[(Int, Long)], RDD[(Array[Int], Array[Int])]) = {
+    val frequentLengthOnePatternAndCounts = getFreqItemAndCounts(minCount, sequences)
```
ditto, I think you can collect directly and use it for filtering.

mengxr commented:
@zhangjiajin Since you already collected the frequent items (length-1 patterns) to driver, you don't need to keep the RDD of length-1 patterns. When generating the final patterns, recomputing the RDD is more expensive than parallelizing the collected ones.

Another comment is to separate local computation from the distributed ones. It makes the implementation easier to read. We can create a private object called `LocalPrefixSpan`, with `run(sequences: Array[Array[Int]], minCount: Int): Array[(Array[Int], Int)]`, then put all local methods under this object.

SparkQA commented:
[Test build #36936 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36936/console) for   PR 7258 at commit [`ba5df34`](https://github.com/apache/spark/commit/ba5df346543e9aee119bd781b257860b65bbe7df).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

zhangjiajin commented:
@mengxr I don't know why method 2 is projection before filtering. I think the method two is exactly what you want. The only need to add functionality to the current code is step 4 (Do we have enough candidates to distribute the work? If no, go to 1 and generate candidates with length + 1.)

![image](https://cloud.githubusercontent.com/assets/13159256/8600512/e4478ea4-2698-11e5-9631-80d26807c03f.png)

![image](https://cloud.githubusercontent.com/assets/13159256/8599886/6fbbca30-2695-11e5-91c4-98272003d303.png)

zhangjiajin committed:
Add new object LocalPrefixSpan, and do some optimization.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37035 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37035/consoleFull) for   PR 7258 at commit [`574e56c`](https://github.com/apache/spark/commit/574e56ccfb271d0ed86c3eba95d1a11a8688495d).

SparkQA commented:
[Test build #37035 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37035/console) for   PR 7258 at commit [`574e56c`](https://github.com/apache/spark/commit/574e56ccfb271d0ed86c3eba95d1a11a8688495d).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/LocalPrefixSpan.scala:
```diff
@@ -0,0 +1,129 @@
+   * @param prefix prefix
+   * @param projectedDatabase the projected dabase
+   * @return a set of sequential pattern pairs,
+   *         the key of pair is pattern (a list of elements),
```
"the key of pair is pattern (a list of elements)," -> "the key of pair is sequential pattern (a list of items),"

zhangjiajin replied:
fixed

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/LocalPrefixSpan.scala:
```diff
@@ -0,0 +1,129 @@
+      maxPatternLength: Int,
+      prefix: Array[Int],
+      projectedDatabase: Array[Array[Int]]): Array[(Array[Int], Long)] = {
+    getPatternsWithPrefix(minCount, maxPatternLength, prefix, projectedDatabase)
```
Why not just rename `getPatternsWithPrefix` -> `run`?

zhangjiajin replied:
fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/LocalPrefixSpan.scala:
```diff
@@ -0,0 +1,129 @@
+private[fpm] object LocalPrefixSpan extends Logging with Serializable {
+
+  /**
+   * Calculate all patterns of a projected database in local.
```
No need for "in local" since object's name makes this clear

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/LocalPrefixSpan.scala:
```diff
@@ -0,0 +1,129 @@
+  }
+
+  /**
+   * Calculate all patterns of a projected database in local.
```
No need for "in local" since object's name makes this clear

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,179 @@
+      sequences: RDD[Array[Int]]): (Array[(Int, Long)], RDD[(Array[Int], Array[Int])]) = {
+    val frequentLengthOnePatternAndCounts = getFreqItemAndCounts(minCount, sequences)
+    val prefixAndProjectedDatabase = getPatternAndProjectedDatabase(
+      frequentLengthOnePatternAndCounts.keys.collect(), sequences)
```
No need to `collect` twice

```scala
val frequentLengthOnePatternAndCounts = getFreqItemAndCounts(minCount, sequences).collect()
val prefixAndProjectedDatabase = getPatternAndProjectedDatabase(
  frequentLengthOnePatternAndCounts.map(_._1), sequences)
(frequentLengthOnePatternAndCounts, prefixAndProjectedDatabase)
```

zhangjiajin replied:
fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,179 @@
+   * @param sequences sequences data
+   * @return prefixes and projected database
+   */
+  private def getPatternAndProjectedDatabase(
```
`getPatternAndProjectedDatabase` -> `getPrefixSuffixPairs`

feynmanliang replied:
Delete `makePrefixProjectedDatabases`, move the `groupByKey()` to here.

zhangjiajin replied:
Because the pair's key is Array(), groupByKey() don't work well, so, the Array must be converted to seq before groupByKey, and be converted back after groupByKey.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,179 @@
+   * Get the frequent prefixes' projected database.
+   * @param frequentPrefixes frequent prefixes
+   * @param sequences sequences data
+   * @return prefixes and projected database
```
"(prefix, suffix) pairs where all prefixes come from frequentPrefixes"

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/fpm/PrefixSpan.scala:
```diff
@@ -0,0 +1,179 @@
+      logWarning("Input data is not cached.")
+    }
+    val minCount = getMinCount(sequences)
+    val (lengthOnePatternsAndCounts, prefixAndCandidates) =
```
Remove `findLengthOnePatterns` and directly set
```scala
val lengthOnePatternsAndCounts = getFreqItemAndCounts(minCount, sequences).collect()
val prefixAndCandidates = getPatternAndProjectedDatabase(
  lengthOnePatternsAndCounts.map(_._1), sequences)
```

zhangjiajin replied:
fixed.

zhangjiajin commented:
@feynmanliang comments: Delete makePrefixProjectedDatabases, move the groupByKey() to the last call in this method (no need to include the two map()s on L161 and L163 since they don't do anything)

Because the pair's key is Array(), groupByKey() don't work well, so, the Array must be converted to seq before groupByKey, and be converted back after groupByKey.

zhangjiajin committed:
Modified the code according to the review comments.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37073 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37073/consoleFull) for   PR 7258 at commit [`ca9c4c8`](https://github.com/apache/spark/commit/ca9c4c8fa84202d8d533c51c277138461ba096a7).

feynmanliang commented:
@zhangjiajin Yep, you're right. Thanks for pointing it out!

feynmanliang commented:
LGTM pending tests

SparkQA commented:
[Test build #37073 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37073/console) for   PR 7258 at commit [`ca9c4c8`](https://github.com/apache/spark/commit/ca9c4c8fa84202d8d533c51c277138461ba096a7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

mengxr commented:
@zhangjiajin Let's merge this version and make improvements in follow-up PRs:

1. `LocalPrefixSpan` performance
  a. `run` should output `Iterator` instead of `Array`.
  b. Local count shouldn't use `groupBy`, which creates too many arrays. We can use `PrimitiveKeyOpenHashMap`.
  c. We can use list to avoid materialize frequent sequences.
2. If there are not enough length-1 patterns to distribute the work, we make another pass to find length-2 patterns.
3. Support non-temporal sequences.

I will make JIRAs for each of them. @feynmanliang will work on 1). @zhangjiajin Could you work on 2) and do some performance benchmark? Thanks!

mengxr commented:
Merged into master. Thanks for contributing PrefixSpan!

zhangjiajin commented:
OK

zhangjiajin committed:
Add feature: Collect enough frequent prefixes before projection in PrefixSpan.

zhangjiajin committed:
fix a scala style error.
